From f1897913939946754a283e46b34889bebef95aeb Mon Sep 17 00:00:00 2001
From: Ash Charles <ashcharles@gmail.com>
Date: Mon, 7 Sep 2015 11:58:34 -0700
Subject: [PATCH 07/24] Add power-off support for the TWL4030 / 6030

Originally proposed by Bernhard Aichriedled and Kris Duff [1] with thanks.

[1] http://gumstix.8.x6.nabble.com/fido-dizzy-gumstix-overo-ironstorm-y-will-not-shutdown-after-a-halt-command-td4970220.html

Omap4 : Fixing the poweroff issues for Duovero

Issue command "poweroff" results in halt. Fix this error and make sure
system is shut down.

Credits to 29d43c2a05c6281d668e9caf95804c3e21c8a35a

Singed-off-by : Jerry Hung <jerry@gumstix.com>
---
 drivers/mfd/Kconfig    |  6 ++++
 drivers/mfd/twl-core.c | 78 ++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 84 insertions(+)

diff --git a/drivers/mfd/Kconfig b/drivers/mfd/Kconfig
index 76f9909..2e238be 100644
--- a/drivers/mfd/Kconfig
+++ b/drivers/mfd/Kconfig
@@ -1572,6 +1572,12 @@ config TWL4030_POWER
 	  and load scripts controlling which resources are switched off/on
 	  or reset when a sleep, wakeup or warm reset event occurs.
 
+config TWL4030_POWEROFF
+	bool "TWL4030/TWL6030 Allow power-off on shutdown"
+	depends on TWL4030_CORE
+	help
+	  Enables the CPU to power-off the system on shutdown
+
 config MFD_TWL4030_AUDIO
 	bool "TI TWL4030 Audio"
 	depends on TWL4030_CORE
diff --git a/drivers/mfd/twl-core.c b/drivers/mfd/twl-core.c
index 104477b..cb1cb3f 100644
--- a/drivers/mfd/twl-core.c
+++ b/drivers/mfd/twl-core.c
@@ -66,6 +66,12 @@
 
 #define DRIVER_NAME			"twl"
 
+#if defined (CONFIG_TWL4030_POWEROFF)
+#define twl_has_poweroff() true
+#else
+#define twl_has_poweroff() false
+#endif
+
 /* Triton Core internal information (BEGIN) */
 
 /* Base Address defns for twl4030_map[] */
@@ -105,6 +111,14 @@
 #define TWL4030_BASEADD_RTC		0x001C
 #define TWL4030_BASEADD_SECURED_REG	0x0000
 
+/* for pm_power_off */
+#define PWR_P1_SW_EVENTS 		0x10
+#define PWR_DEVOFF	 		(1 << 0)
+
+#define TWL6030_PHOENIX_DEV_ON	0x25
+#define TWL6030_APP_DEVOFF	(1<<0)
+#define TWL6030_CON_DEVOFF	(1<<1)
+#define TWL6030_MOD_DEVOFF	(1<<2)
 /* Triton Core internal information (END) */
 
 
@@ -1071,11 +1085,66 @@ static int twl_remove(struct i2c_client *client)
 	return 0;
 }
 
+static void twl_4030_poweroff(void)
+{
+	int err;
+	u8 val;
+
+	err = twl_i2c_read_u8(TWL_MODULE_PM_MASTER, &val,
+				  PWR_P1_SW_EVENTS);
+	if (err) {
+		pr_err("%s: i2c error %d while reading TWL4030"
+			"PM_MASTER P1_SW_EVENTS\n",
+			DRIVER_NAME, err);
+		return;
+	}
+
+	val |= PWR_DEVOFF;
+
+	err = twl_i2c_write_u8(TWL_MODULE_PM_MASTER, val,
+				   PWR_P1_SW_EVENTS);
+	if (err)
+		pr_err("%s: i2c error %d while writing TWL4030"
+			"PM_MASTER P1_SW_EVENTS\n",
+			DRIVER_NAME, err);
+}
+
+
 static struct of_dev_auxdata twl_auxdata_lookup[] = {
 	OF_DEV_AUXDATA("ti,twl4030-gpio", 0, "twl4030-gpio", NULL),
 	{ /* sentinel */ },
 };
 
+static void twl_6030_poweroff(void)
+{
+	int err;
+	u8 val;
+
+	err = twl_i2c_read_u8(TWL6030_MODULE_ID0, &val,
+	                      TWL6030_PHOENIX_DEV_ON);
+	if (err) {
+		pr_err("%s: i2c error %d while reading TWL6030"
+		       "MODULE_ID0 PHOENIX_DEV_ON\n",
+		       DRIVER_NAME, err);
+		return;
+	}
+
+	val |= TWL6030_APP_DEVOFF;
+	val |= TWL6030_CON_DEVOFF;
+	val |= TWL6030_MOD_DEVOFF;
+
+	err = twl_i2c_write_u8(TWL6030_MODULE_ID0, val,
+	                       TWL6030_PHOENIX_DEV_ON);
+
+	if (err) {
+		pr_err("%s: i2c error %d while writing TWL6030"
+		       "MODULE_ID0 PHOENIX_DEV_ON\n",
+		       DRIVER_NAME, err);
+	}
+
+	return;
+}
+
 /* NOTE: This driver only handles a single twl4030/tps659x0 chip */
 static int
 twl_probe(struct i2c_client *client, const struct i2c_device_id *id)
@@ -1226,6 +1295,15 @@ twl_probe(struct i2c_client *client, const struct i2c_device_id *id)
 				 TWL4030_DCDC_GLOBAL_CFG);
 	}
 
+	if (twl_has_poweroff())
+	{
+		/* initialize pm_power_off routine */
+		if (twl_class_is_6030())
+			pm_power_off = twl_6030_poweroff;
+		else
+			pm_power_off = twl_4030_poweroff;
+	}
+
 	if (node) {
 		if (pdata)
 			twl_auxdata_lookup[0].platform_data = pdata->gpio;
-- 
2.7.4

